### ISSUE TITLE
DataGrid and DataTable: rows- and first-attributes ignored after use of paginator

### ISSUE THREAD

**Issue by matinh** [primefaces/primefaces#4931]
## 1) Environment
- PrimeFaces version: 7.0/7.1-SNAPSHOT
- Does it work on the newest released PrimeFaces version? Version? No, neither with 7.0, nor with 7.1-SNAPSHOT.
- Does it work on the newest sources in GitHub? (Build by source -> https://github.com/primefaces/primefaces/wiki/Building-From-Source). No.
- Application server + version: Tested with Wildfly 11 and Jetty 9.4.15 from the primefaces-test project.
- Affected browsers: probably all (tested with Firefox 60.7.0esr, Opera 58 and Chromium 75). Note: Opera and Chromium seem to have other problems with 7.1-SNAPSHOT, which is probably unrelated)

## 2) Expected behavior

The `rows`-attribute from the `<p:dataGrid>` component should always use the actual value.

## 3) Actual behavior

After the use of the pager, the `rows`-attribute seems to be ignored and the last value before the pager was used is always used, no matter what the actual value of the rows EL is.

## 4) Steps to reproduce

* clone https://github.com/matinh/primefaces-test.git using the branch _dataGrid_ with `git clone 'https://github.com/matinh/primefaces-test.git' -b dataGrid`
* change directory `cd primefaces-test`
* run `mvn jetty:run-exploded`
* goto http://localhost:8080/primefaces-test/
* use the spinner to change number of rows in the dataGrid
* use pager to switch to another page
* try to use spinner again to change number of rows, which is no longer working.

## Further comments

* There are no JavaScript Errors in the browsers console.
* The setter of the columns-attribute is called with the correct value.

**Comment by Rawi01** [comment]
I checked some primefaces sources and it seems like the problem is in L297 https://github.com/primefaces/primefaces/blob/e2062dc5671b10e9f3913b454d4c31b86a62c226/src/main/java/org/primefaces/component/api/UIData.java#L290-L307

A possible fix might be to check for an exisitng value expression as in L305 and only set that value to the new one e.g.

```java
ValueExpression rowsVe = data.getValueExpression("rows"); 

if (rowsVe != null && !rowsVe.isReadOnly(elContext)) { 
    rowsVe.setValue(context.getELContext(), Integer.valueOf(rowsParam)); 
} else {
    data.setRows(Integer.valueOf(rowsParam));
}
```

**Comment by matinh** [comment]
Thanks @Rawi01 for your comment. I tested your suggestion and it fixes the problem for me. I created PR #4937 as a fix. BTW, the issue is not only with `<p:dataGrid>` but also with `<p:dataTable>`. The suggested change fixes the issue for both components.

One more thing: from looking at the source, the same issue might arise with the `first`-attribute. Should we change this in a similar manner as well?

**Comment by Rawi01** [comment]
Thanks for testing my idea. I think that the `first` attribute should be handled the same. We should consider to move this functionality into a utility method.

**Comment by matinh** [comment]
I just verified with the test-project, that first has the same problem. I'll update the PR to also fix this issue.

**Comment by matinh** [comment]
Ok, I've updated the PR #4937:
* it now considers the columns attribute and the first attribute
* it removes some unnecessary code which updates the attributes again.
* it refactors the check for a writable value-expression into a method `isWriteable()`

I'm relatively new to git and messed up my branch for this fix. I did undo some accidental commits. The commits containing the actual changes are 09ffdd9 and 2b8963a. I hope the PR is useable. If not, let me know and I'll try to create a new one.

**Comment by matinh** [comment]
@Rawi01: @tandraschko won't apply the PR until you explain why your suggestion fixes the issue. See https://github.com/primefaces/primefaces/pull/4937#issuecomment-505469850 for details. Could you please elaborate? Thanks.

**Comment by Rawi01** [comment]
@matinh @tandraschko I had a few free minutes, looked at the sources and found somehting that seems to be the problem. I never tested or debugged it, it was meant to be a starting point for someone who wants to fix this bug.

About the fix: I assumed that the whole bug report is right at the beginning. That means that I know that it works before the paginator gets used and the problem must be related to the paginator. I looked up the place where the rows gets set by the paginator and found the lines above. I noticed that it writes the value two times , once using `setRows` and once to the value expression. I was courious why that breaks it and found out that the state helper that stores the value and is used in `getRows` tries to use a static value first and only if it is not present it tries to evaluate a value expression. Thats why I suggested to fix it by only writing the value to the value expression as this would not break the two way binding (not sure if it is called that way in jsf) by adding an additional static value.

**Comment by matinh** [comment]
Ok, I did some further research to see what actually happens. Here we go:
The [`DataGridRenderer`](https://github.com/primefaces/primefaces/blob/master/src/main/java/org/primefaces/component/datagrid/DataGridRenderer.java) when rendering the grid, only uses `UIData.getRows()` and never queries the ValueExpression directly. 

`UIData.getRows()` is like this:
```java
    return (Integer) getStateHelper().eval(PropertyKeys.rows, 0);
```
and `eval()` is defined like this:

```java
    public Object eval(Serializable key, Object defaultValue) {
        Object retVal = get(key);
        if (retVal == null) {
            ValueExpression ve = component.getValueExpression(key.toString());
            if (ve != null) {
                retVal = ve.getValue(component.getFacesContext().getELContext());
            }
        }
        return ((retVal != null) ? retVal : defaultValue);
    }
```
So as long as `setRows()` was not called, `getRows()` effectively uses `component.getValueExpression('rows')` which is what we want.

When the pager is used, [`updatePagination()`](https://github.com/primefaces/primefaces/blob/e2062dc5671b10e9f3913b454d4c31b86a62c226/src/main/java/org/primefaces/component/api/UIData.java#L283) is called which, in the unmodified version, sets the new rows-value to whtaever was submitted by the pager:
https://github.com/primefaces/primefaces/blob/e2062dc5671b10e9f3913b454d4c31b86a62c226/src/main/java/org/primefaces/component/api/UIData.java#L283-L297

From now on, calls to `UIData.getRows()` return the last value submitted by the pager and never use the ValueExpression like before.

The changes from PR https://github.com/primefaces/primefaces/pull/4937 now avoid using the setter when a ValueExpression was used for the corresponding attributes but instead write to the ValueExpression. Only when no (writeable) VE was in use the submitted value is written directly into the component state.

Conclusion: IMHO the suggested fix should be ok. I'm not 100% sure if this is the suggested way, how PrimeFaces components should handle this situation, but currently I do not see any problems with this approach and I did quite some testing with DataGrid, and DataTable, included filtering, sorting, mutiViewState, paging, etc.
